File: python/063. Unique Paths II.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m=len(obstacleGrid)
        n=len(obstacleGrid[0])
        matrix= [([0] * n) for i in range(m)]
        print 'matrix[0][0]='+str(matrix[0][0])
        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j]==1:
                    matrix[i][j]=0
                elif i==0 and j==0:
                    matrix[0][0]=1-obstacleGrid[0][0]
                else:
                    #负数是反向读第一个，为0，不影响结果
                    matrix[i][j]=matrix[i-1][j]+matrix[i][j-1]
        return matrix[m-1][n-1]

File: java/063. Unique Paths II.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
    	if (obstacleGrid[0][0]==1||obstacleGrid[obstacleGrid.length-1][obstacleGrid[0].length-1]==1) {
			return 0;
		}
    	Map<String,Integer> map=new HashMap<String, Integer>();
    	return f(obstacleGrid,0,0,obstacleGrid.length-1,obstacleGrid[0].length-1,map);
    }
    public int f(int[][] obstacleGrid,int x,int y,int xl,int yl,Map<String,Integer> map){
    	String coordinate=x+","+y;
    	if (x==xl&&y==yl) {
			return 1;
		}
    	if (map.get(coordinate)==null) {
			boolean xflag=(x<xl)&&(obstacleGrid[x+1][y]==0);
			boolean yflag=(y<yl)&&(obstacleGrid[x][y+1]==0);
			int value=(xflag?f(obstacleGrid,x+1,y,xl,yl,map):0)+(yflag?f(obstacleGrid,x,y+1,xl,yl,map):0);
			map.put(coordinate, value);
			return value;
		}
    	return map.get(coordinate);
    }
}